snuba/web/query.py

```diff
@@ -164,15 +165,21 @@ def _run_query_pipeline(
     else:
         query_runner = partial(
             _run_and_apply_column_names,
-            timer,
-            query_metadata,
-            request.referrer,
-            robust,
-            concurrent_queries_gauge,
+            timer=timer,
+            query_metadata=query_metadata,
+            attribution_info=request.attribution_info,
+            robust=robust,
+            concurrent_queries_gauge=concurrent_queries_gauge,
         )
 
     record_missing_tenant_ids(request)
+    # NOTE: Something which is not immediately clear from looking at the code is that the
+    # query_runner can be run multiple times during the execution of the pipeline.
+    # The execution pipeline may choose to break up a query into multiple subqueries. And
+    # then assemble those together into one resut
 
+    # Throughout those executions, the query_metadata.query_list is appended to every time a query runs
```

Comment:
Nice! One question I have is: are there any differences between passing `query_metadata. query_list` by reference into `raw_query()` and passing the entire object when it comes to safety with multiple executions? Do we parallelize this process?

Reply:
Good spot! The answer to your question is Yes and No.

There is the [MultipleConcurrentPipeline](https://github.com/getsentry/snuba/blob/master/snuba/pipeline/pipeline_delegator.py#L73) which takes a `runner` callable and runs different versions of the pipeline in different threads. In the case that is used, this would indeed be a race condition.

Currently [`SnubaQueryMetadata`](https://github.com/getsentry/snuba/blob/master/snuba/querylog/query_metadata.py#L209) does not do any sort of locking on the `query_list` so there is nothing preventing this race condition.

All that being said, the `MultipleConcurrentPipeline` is not actually used anywhere currently but even when it was, we did have this problem. Also, python threads are not truly parallel so we are fine on this front overall

Reply:
I updated the comment to provide more info. Let me know if that makes sense

Reply:
Cool, thanks for the clarification and the added comment 👍